Dev Gfx Reg Helper

jiaxie-amd opened the pull request:
- Add GfxRegHelper Class
- Add SqImgSampRegHelper Class for Gfx9 and Gfx10
- Add SqImgRsrcRegHelper Class
- Support SqImgRsrcRegHelper for Gfx10

amdvlk-admin commented:
### Test summary for commit 4f40312f68c9fd49a526730cd8f1ed7ba131b16c
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev bb7398b1c7509a30ba4f7d947b5ad5549a4d58db</li>
<li>SPVGEN: dev b30c124e7a042c5251aeb20780a83089355e74a1</li>
<li>XGL: dev ffd39443fec0ce821b7702173eba7184855c7b5e</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ebc431ac299fa294a4e730432520d37f326843c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103137/103137 (100.0%)</li>
<li>Failed: 0/103137 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103193/103193 (100.0%)</li>
<li>Failed: 0/103193 (0.0%)</li>
</ul></details>


</ul></details>

kuhar commented on llpc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,314 @@
+        m_bitsInfo = g_sqImgRsrcRegBitsGfx10;
+        break;
+    default:
+        llvm_unreachable("GfxIp is not supported!");
```
maybe just: `assert((version == 9 || version == 10) && "...")?

kuhar commented on llpc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,139 @@
+};
+
+// SqImgRsrcRegisters Id
+enum SqRsrcRegs
```
Since this is in a header, could you make it into an `enum class`?

Where do these values come from? Is there a relevant piece of spec we could reference in the comment?

jiaxie-amd replied:
Please refer to https://github.com/GPUOpen-Drivers/llpc/pull/527

kuhar commented on llpc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,139 @@
+};
+
+// SqImgSampRegisters Id
+enum SqSampRegs
```
Same as above.

kuhar commented on llpc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,139 @@
+    Value* GetRegCommon(uint32_t id);
+
+    // Set register
+    inline void SetRegCommon(uint32_t id, Value* pVal)
```
nit: `inline` is automatically applied to in-class method definitions and could be omitted here.

kuhar commented on llpc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,139 @@
+};
+
+// =====================================================================================================================
+// Helper class for handling graphics registers
```
Could you briefly mention what these classes help with? Based on the comments and class definitions, it's unclear to me what these classes do and why.

kuhar commented on llpc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,314 @@
+// SqImgSampReg Bits infomation look up table (Gfx9)
+static constexpr BitsInfo g_sqImgRsrcRegBitsGfx9[SqRsrcRegsCount] =
+{
+    { 0,  0, 32 }, // BaseAddress
```
Is there relevant docs/specs that mention these values?

jiaxie-amd replied:
Added comments

kuhar commented on llpc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,314 @@
+// SqImgSampReg Bits infomation look up table (Gfx10)
+static constexpr BitsInfo g_sqImgRsrcRegBitsGfx10[SqRsrcRegsCount] =
+{
+    { 0,  0, 32 }, // BaseAddress
```
Same.

jiaxie-amd replied:
Added comments

jiaxie-amd commented:
> Could you give a high level description of what the code is trying to achieve and how it connects with the rest of the project?

Yeah, I've wrote some description in https://github.com/GPUOpen-Drivers/llpc/pull/527, please take a look.

jiaxie-amd commented on llpc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,139 @@
+    void SetRegCombine(uint32_t idLo, uint32_t idHi, Value* pReg);
+
+protected:
+    BitsState*      m_bitsState;
```
Okay, I'd use unique_ptr for compatibility for now.

jiaxie-amd commented on llpc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,139 @@
+
+protected:
+    BitsState*      m_bitsState;
+    const BitsInfo* m_bitsInfo;
```
Revised

jiaxie-amd commented on llpc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,139 @@
+namespace lgc
+{
+
+using namespace llvm;
```
Thanks for reminding, but this issue exists in other files too, maybe we appliy the modification in the whole project with one change.

@trenouf, will you prepare to refine this latter?

jiaxie-amd commented on llpc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,139 @@
+
+using namespace llvm;
+
+struct BitsState
```
Added comments.

jiaxie-amd commented on llpc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,314 @@
+}
+
+// =====================================================================================================================
+// Set data into two seperate DWORDs
```
Please refer to https://github.com/GPUOpen-Drivers/llpc/pull/527

jiaxie-amd commented on llpc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,314 @@
+    m_bitsState     = nullptr;
+}
+
+GfxRegHelper::~GfxRegHelper()
```
Removed since we are now using unique ptr

trenouf commented on llpc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,111 @@
+/*
+ ***********************************************************************************************************************
+ *
+ *  Copyright (c) 2019-2020 Advanced Micro Devices, Inc. All Rights Reserved.
```
I believe we should only put one year in the copyright message, the year of creation.

jiaxie-amd replied:
Updated to 2020 only

trenouf commented:
> > Could you give a high level description of what the code is trying to achieve and how it connects with the rest of the project?
> 
> Yeah, I've wrote some description in #527, please take a look.

Could we please have some documentation for the class in a comment in the code, instead of (or as well as) the document?

jiaxie-amd commented:
> I have just merged PR504 that moves LGC into its own lgc directory. Please rebase, and note that your new files now belong in lgc/util not llpc/util.

This change is rebased on PR504, but it seems the Makefile.llpc is not fully updated? Or is there a Makefile.lgc

amdvlk-admin commented:
### Test summary for commit 248f4a81fa8dc55c68c6b9040a58621a61a4474a
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev bb7398b1c7509a30ba4f7d947b5ad5549a4d58db</li>
<li>SPVGEN: dev b30c124e7a042c5251aeb20780a83089355e74a1</li>
<li>XGL: dev ffd39443fec0ce821b7702173eba7184855c7b5e</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ebc431ac299fa294a4e730432520d37f326843c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103137/103137 (100.0%)</li>
<li>Failed: 0/103137 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103193/103193 (100.0%)</li>
<li>Failed: 0/103193 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 811146db6e8ea8acae531343992f18d38bf07003
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev bb7398b1c7509a30ba4f7d947b5ad5549a4d58db</li>
<li>SPVGEN: dev b30c124e7a042c5251aeb20780a83089355e74a1</li>
<li>XGL: dev ffd39443fec0ce821b7702173eba7184855c7b5e</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ebc431ac299fa294a4e730432520d37f326843c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103137/103137 (100.0%)</li>
<li>Failed: 0/103137 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103193/103193 (100.0%)</li>
<li>Failed: 0/103193 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit f65cddbd1400d81e59c1de5949bcc38dd98bf668
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev bb7398b1c7509a30ba4f7d947b5ad5549a4d58db</li>
<li>SPVGEN: dev b30c124e7a042c5251aeb20780a83089355e74a1</li>
<li>XGL: dev ffd39443fec0ce821b7702173eba7184855c7b5e</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ebc431ac299fa294a4e730432520d37f326843c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103137/103137 (100.0%)</li>
<li>Failed: 0/103137 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103193/103193 (100.0%)</li>
<li>Failed: 0/103193 (0.0%)</li>
</ul></details>


</ul></details>

kuhar commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,145 @@
+ /**
+  ***********************************************************************************************************************
+  * @file  llpcGfxRegHelper.h
+  * @brief LLPC header file: contains the definition of LLPC utility class GfxRegHelper.
```
Can this comment be more descriptive? This is basically the same as the `@file` line above.

I think it would help a lot to give an introduction that mentions what the high-level goal is and reference the doc from the other PR.

jiaxie-amd replied:
Yeah, this description is meaningless, but i had to follow this format aligned with the other files in this project :(, I would provide the link to the doc

jiaxie-amd replied:
I'd add the path of the document to this description once https://github.com/GPUOpen-Drivers/llpc/pull/527 is being merged.

kuhar replied:
Sounds good to me.

kuhar commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,145 @@
+// =====================================================================================================================
+// Struct maintains state for specific register
+// - pValue : Cache the value of the register which is being fetched already.
+// - isModified : Set to be true if the register is being settled with new value.
```
It's unclear to me what 'state' means here. Is it a symbolic expression over some pre-defined graphics register, or register as in arbitrary SSA value?
What does it mean for a register to be settled with a value? Do you write to a hardware register, construct a new llvm value over a previously read register value?

jiaxie-amd replied:
e.g. if we want to get A2, and may keep the A2 if we did do some modification on it, and may want to use it in someplace else. 
```
D0 = extractelement(0) // DWORD0
M0 = 0x1111                 // mask
A1 =  M0 & D0              // specific register
A2 = A1 + 1                   // do some modification
D1 =(D0 & ~M0) | (A2 & M0) // update DWORD0
insertelement(0)                     // insert back
```

kuhar replied:
I'd say that these are just symbolic expressions over values read from hardware graphics registers, materialized as llvm instructions. The code maintains a map from from a register ID to an llvm value with the current symbolic expression describing the updated state of the register.

The goal of the helper structs is to maintain the newest symbolic state of a register and eventually produce optimized llvm instruction sequences for that states.

If that sounds reasonable, could you add a similar description in comments?

jiaxie-amd replied:
Thanks, i'll update the comment as you suggested.

kuhar commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,145 @@
+};
+
+// =====================================================================================================================
+// Helper class for handling graphics registers
```
Are these hardware registers, llvm values with known names that correspond to amdgpu registers, or something else?

jiaxie-amd replied:
These are hardware registers.

kuhar replied:
Fantastic, can you update the comment to mention that if that's not already in the top-level comment?

jiaxie-amd replied:
These are mentioned in the top level comment now.

kuhar commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,145 @@
+    virtual ~GfxRegHelper() = default;
+
+    // Get register
+    Value* GetRegCommon(uint32_t id);
```
Where do the ids come from? What are the possible values for `id`?

jiaxie-amd replied:
The id is the index of BitsInfo and BitsState
```
// =====================================================================================================================
// SqImgSampReg Bits infomation look up table (Gfx9-10)
// Refer to imported/chip/gfx9/gfx9_plus_merged_registers.h : SQ_IMG_SAMP_WORD
static constexpr BitsInfo g_sqImgSampRegBitsGfx9[static_cast<uint32_t>(SqSampRegs::SqSampRegsCount)] =
{
    { 0, 30, 2 }, // FilterMode
    { 2, 20, 2 }, // XyMagFilter
    { 2, 22, 2 }, // XyMinFilter
};

// SqImgSampRegisters Id
// Corresponding to g_sqImgSampRegBitsGfx9
enum class SqSampRegs
{
    FilterMode = 0,
    xyMagFilter,
    xyMinFilter,

    SqSampRegsCount,
};
```

jiaxie-amd replied:
Acutally, we don't expect user to use `GfxRegHelper` directly, but only implement its own register helper class, such as `SqImgSampRegHelper`

kuhar replied:
Sounds good to me. Can we have it in the comments?

jiaxie-amd replied:
Updated.

kuhar commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    // Get combined data from two seperate DWORDs
+    Value* GetRegCombine(uint32_t idLo, uint32_t idHi);
```
What is the llvm type of the returned value? I imagine that it'd be a 2-dword-large integer type, or maybe a vector type?

jiaxie-amd replied:
The return time is a dword, it doesn't support 2-dword for now.
```
Dword 0
{
    unused                : 30 
    width_low            :  2
}

Dword 1
{
    width_hi               :  12
    unused                :  20 
}
```
I had considered the naming of this function and there should be some restirctions to the low and high BitsInfo. But, there's no usage that really need a 2-dword output value.

kuhar replied:
I think this is important and should be included in the comment. I agree that finding a good name is difficult here.

jiaxie-amd replied:
Yes, i'll comment this information.

kuhar commented on lgc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,308 @@
+        llvm_unreachable("GfxIp is not supported!");
+        break;
+    }
+    m_bitsState.reset(new BitsState[static_cast<uint32_t>(SqSampRegs::SqSampRegsCount)]);
```
nit: you can use `std::make_unique` here

```
m_bitsState = std::make_unique<BitsState[]>(static_cast<uint32_t>(SqSampRegs::SqSampRegsCount));
```

kuhar commented on lgc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,308 @@
+    //  - The register has not being initialized
+    //  - The register is being modified
+    if ((m_bitsState[id].pValue == nullptr) ||
+        (m_bitsState[id].isModified == 1))
```
nit: `isModified` is a bool, so you can use it directly instead of doing the comparison

amdvlk-admin commented:
### Test summary for commit 4afa4e3302ed64dbfea7db1bc704cee454f3fac5
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev bb7398b1c7509a30ba4f7d947b5ad5549a4d58db</li>
<li>SPVGEN: dev b30c124e7a042c5251aeb20780a83089355e74a1</li>
<li>XGL: dev ffd39443fec0ce821b7702173eba7184855c7b5e</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ebc431ac299fa294a4e730432520d37f326843c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103137/103137 (100.0%)</li>
<li>Failed: 0/103137 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103193/103193 (100.0%)</li>
<li>Failed: 0/103193 (0.0%)</li>
</ul></details>


</ul></details>

kuhar commented on llpc/make/Makefile.llpc:
```diff
@@ -231,4 +231,3 @@ else
 endif
 
 .FORCE:
-
```
nit: unintended change?

jiaxie-amd replied:
I think this is modified by my editoer, i'll rever it back to the original version.

kuhar commented on lgc/CMakeLists.txt:
```diff
@@ -230,9 +230,9 @@ target_sources(LLVMlgc PRIVATE
     util/llpcPipelineShaders.cpp
     util/llpcStartStopTimer.cpp
 )
-
```
nit: unintended change?

amdrexu commented:
Once this change is logically finalized, please ping me. I will just have a coding check. Also, please click "Resolve Conversation" if the review issues are addressed. Thanks very much.

jiaxie-amd commented:
Hi Rex,

Okay, by the way I hope https://github.com/GPUOpen-Drivers/llpc/pull/527 could be merged first, at least, should be agreed to be merged, then I could add the link of this doc to this PR525.

Please let me know if any modification needs to be taken(especially file name, and file location, I’ll reference the doc)

Best Regards,
Harris

From: Rex Xu <notifications@github.com>
Sent: Tuesday, March 31, 2020 10:52
To: GPUOpen-Drivers/llpc <llpc@noreply.github.com>
Cc: Xie, Jiajun(Harris) <Jiajun.Xie@amd.com>; Author <author@noreply.github.com>
Subject: Re: [GPUOpen-Drivers/llpc] Dev Gfx Reg Helper (#525)


Once this change is logically finalized, please ping me. I will just have a coding check. Also, please click "Resolve Conversation" if the review issues are addressed. Thanks very much.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FGPUOpen-Drivers%2Fllpc%2Fpull%2F525%23issuecomment-606367433&data=02%7C01%7Cjiajun.xie%40amd.com%7C5d75a1f2693943fb088e08d7d51e8a76%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C637212199454744155&sdata=Lu0Yh4RdGx2rTN0aEiEJz3sMqEffijhsZlljWZ7Fehc%3D&reserved=0>, or unsubscribe<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAN3W45OQ7GF3JC6ZYUUGCPDRKFLGPANCNFSM4LVBKMAA&data=02%7C01%7Cjiajun.xie%40amd.com%7C5d75a1f2693943fb088e08d7d51e8a76%7C3dd8961fe4884e608e11a82d994e183d%7C0%7C0%7C637212199454744155&sdata=DUseMYKFTZ4NU5OmCuBdtZAh4AMBFOUAJHZIfO%2FtAoQ%3D&reserved=0>.

jiaxie-amd commented:
> Once this change is logically finalized, please ping me. I will just have a coding check. Also, please click "Resolve Conversation" if the review issues are addressed. Thanks very much.

Hi Rex, could you please take a look at this change, thanks!

amdvlk-admin commented:
### Test summary for commit dd9a7336eec89f3327187ab023e3eff81a2ae2b4
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev bb7398b1c7509a30ba4f7d947b5ad5549a4d58db</li>
<li>SPVGEN: dev b30c124e7a042c5251aeb20780a83089355e74a1</li>
<li>XGL: dev ffd39443fec0ce821b7702173eba7184855c7b5e</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ebc431ac299fa294a4e730432520d37f326843c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103137/103137 (100.0%)</li>
<li>Failed: 0/103137 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103193/103193 (100.0%)</li>
<li>Failed: 0/103193 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 4b5cc1b0bc611b411b1faf1e8d8230efc5ebc775
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev bb7398b1c7509a30ba4f7d947b5ad5549a4d58db</li>
<li>SPVGEN: dev b30c124e7a042c5251aeb20780a83089355e74a1</li>
<li>XGL: dev ffd39443fec0ce821b7702173eba7184855c7b5e</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ebc431ac299fa294a4e730432520d37f326843c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103137/103137 (100.0%)</li>
<li>Failed: 0/103137 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103193/103193 (100.0%)</li>
<li>Failed: 0/103193 (0.0%)</li>
</ul></details>


</ul></details>

kuhar commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,153 @@
+// The goal of the helper structs is to maintain the newest symbolic state of a register and eventually produce
+// optimized llvm instruction sequences for these states.
+// - pValue : Cache the value which is already being fetched.
+// - isModified : True, if the register is being settled with new value.
```
In general, I don't think you can say that 'a register is settled with a value'. I may be wrong here, but it's not a standard term from what I know, and it'd be better to define it directly.
Maybe say instead that: 'This indicates that the value read from a hardware register was given a new symbolic state (on top of a freshly read register value). Note that we don't actually modify the state of hardware registers.

jiaxie-amd replied:
I just droped the previous comments and added a state table on it.

kuhar commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,153 @@
+    virtual ~GfxRegHelper() = default;
+
+    // Get register
+    Value* GetRegCommon(uint32_t id);
```
This hasn't been resolved. Can you add a quick comment that explains where the ID can come from?
This can be a comment on top of the class instead, since it covers many methods that all take IDs.

kuhar replied:
For all the subclasses, if the only valid IDs come from known enums, maybe make this protected and expose by subclass methods that take enums instead of uint32?

jiaxie-amd replied:
> This hasn't been resolved. Can you add a quick comment that explains where the ID can come from?
> This can be a comment on top of the class instead, since it covers many methods that all take IDs.

I've added comment on the top of the class.

kuhar commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,153 @@
+};
+
+// SqImgSampRegisters Id
+// Corresponding to g_sqImgSampRegBitsGfx9
```
nit: Corresponds

kuhar commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,153 @@
+    SqImgSampRegHelper(Builder* pBuilder, Value* pRegister, GfxIpVersion* pGfxIpVersion);
+
+    // Acquire register value
+    Value* GetReg(SqSampRegs id);
```
If this is a virtual function, can you mark it with `override` to make it clear? Same with all the other virtual functions here and the other subclasses that override functions.

jiaxie-amd replied:
No, this is not a virtual function, and there should be no other virtual functions.

kuhar commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,153 @@
+public:
+    SqImgSampRegHelper(Builder* pBuilder, Value* pRegister, GfxIpVersion* pGfxIpVersion);
+
+    // Acquire register value
```
The previous class documented this as: `Get register`.
I think it'd be better to be consistent and say something like `Get the current llvm value for the hardware register`.
What do you think? Similar with set.

jiaxie-amd replied:
I've update the comment with your suggestion, but left he function name unmodified since its sufficient to present the meaning and there should be no conflictions.

kuhar commented on lgc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,311 @@
+
+// =====================================================================================================================
+// SqImgSampReg Bits infomation look up table (Gfx10)
+// Refer to: (Not released)
```
Can you add a `TODO` or `FIXME` to fill it in later?

jiaxie-amd replied:
Added TODO

kuhar commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,153 @@
+    Value*                       m_pInt32One     = nullptr;
+    GfxIpVersion*                m_pGfxIpVersion = nullptr;
+    const BitsInfo*              m_bitsInfo      = nullptr;
+    std::unique_ptr<BitsState[]> m_bitsState     = nullptr;
```
Can we have a comment that says that this is a map from register ID (?) to its current state? And because the number of IDs is small, it's implemented with an array.

jiaxie-amd replied:
Comment on the top of the class now.

nhaehnle commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,153 @@
+protected:
+    GfxRegHelper(Builder* pBuilder, Value* pRegister);
+
+    virtual ~GfxRegHelper() = default;
```
You can remove this: there is no need for a virtual destructor here, since the class is made to behave very much value-like.

jiaxie-amd replied:
Yes, this is unnecessary now.

nhaehnle commented on lgc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,311 @@
+        llvm_unreachable("GfxIp is not supported!");
+        break;
+    }
+    m_bitsState = std::make_unique<BitsState[]>(static_cast<uint32_t>(SqRsrcRegs::SqRsrcRegsCount));
```
Actually, I spoke a bit too soon. Is there a way to avoid this dynamic allocation here? Since the size is constant, it should be possible to embed the array directly into the class.

jiaxie-amd replied:
Actually I have written a version that use stack rather than new something else. But just considering if the bitsinfo are autogenerated, then maybe the state is too large.
Anyway, it is easy to eliminate the usage of new here, and since the current version contains few bitstates, just be fine.

jiaxie-amd replied:
The current version is now without new operation, please review, thanks.

jiaxie-amd commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,153 @@
+// =====================================================================================================================
+// The goal of the helper structs is to maintain the newest symbolic state of a register and eventually produce
+// optimized llvm instruction sequences for these states.
+// - pValue : Cache the value which is already being fetched.
```
I've updated the comments, please help review.

jiaxie-amd commented:
> I have just merged PR504 that moves LGC into its own lgc directory. Please rebase, and note that your new files now belong in lgc/util not llpc/util.

I think the code files are now in proper location, thanks.

amdvlk-admin commented:
### Test summary for commit d4da446341c03292cdb7e66ee0ebe1f03993165d
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev bb7398b1c7509a30ba4f7d947b5ad5549a4d58db</li>
<li>SPVGEN: dev b30c124e7a042c5251aeb20780a83089355e74a1</li>
<li>XGL: dev ffd39443fec0ce821b7702173eba7184855c7b5e</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1ebc431ac299fa294a4e730432520d37f326843c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103137/103137 (100.0%)</li>
<li>Failed: 0/103137 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103193/103193 (100.0%)</li>
<li>Failed: 0/103193 (0.0%)</li>
</ul></details>


</ul></details>

jiaxie-amd commented:
@amdrexu @amdvlk-admin , 
I've checked the jenkins result:
Test run totals:
  Passed:        63878/103137 (61.9%)
  Failed:        0/103137 (0.0%)
  Not supported: 39259/103137 (38.1%)
  Warnings:      0/103137 (0.0%)
No failure found.
Could you please check if there's some issue on the building system, thanks.

amdrexu commented on lgc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,311 @@
+        m_bitsInfo = g_sqImgRsrcRegBitsGfx10;
+        break;
+    default:
+        llvm_unreachable("GfxIp is not supported!");
```
GfxIp -> GFX IP

amdrexu commented on lgc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,311 @@
+    case SqRsrcRegs::IsTileOpt:
+        return m_pBuilder->CreateICmpNE(GetRegCommon(static_cast<uint32_t>(id)), m_pBuilder->getInt32(0));
+    default:
+        llvm_unreachable("Not implemented!");
```
The same as above
// TODO: More will be implemented.
llvm_unreachable("Not implemented!");
or
llvm_unreachable("Invalid register ID!");

amdrexu commented on lgc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,311 @@
+// =====================================================================================================================
+// Set the current llvm value for the hardware register
+void SqImgRsrcRegHelper::SetReg(
+    SqRsrcRegs id,     // Register ID
```
id -> regId

amdrexu commented on lgc/util/llpcGfxRegHelper.cpp:
```diff
@@ -0,0 +1,311 @@
+// Set the current llvm value for the hardware register
+void SqImgRsrcRegHelper::SetReg(
+    SqRsrcRegs id,     // Register ID
+    Value*     pParam) // [in] Param to be set
```
pParam -> pRegValue // [in] Value to set to this register

amdrexu commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,181 @@
+ **********************************************************************************************************************/
+ /**
+  ***********************************************************************************************************************
+  * @file  llpcGfxRegHelper.h
```
llpcGfxRegHelper.h -> llpcGfxRegHandler.h

amdrexu commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,181 @@
+//         xyMagFilter,         <==        { 2, 20, 2 }, // XyMagFilter    ==>      {nullptr, false},
+//         xyMinFilter,                    { 2, 22, 2 }, // XyMinFilter             {nullptr, false},
+//       };                              };                                       };
+class GfxRegHelper : public GfxRegHandlerBase
```
GfxRegHelper -> GfxRegHandler

amdrexu commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,181 @@
+    void SetBitsState(BitsState* bitsState) { m_bitsState = bitsState; }
+
+protected:
+    Value*          m_pInt32One     = nullptr;
```
m_pInt32One -> m_pOne = nullptr; // Int32 constant one
Since in this file, there is only int32 one. No need of adding the type prefix "Int32".

amdrexu commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,181 @@
+protected:
+    Value*          m_pInt32One     = nullptr;
+    GfxIpVersion*   m_pGfxIpVersion = nullptr;
+    const BitsInfo* m_bitsInfo      = nullptr;
```
m_bitsInfo  -> m_pBitsInfo

amdrexu commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,181 @@
+    const BitsInfo* m_bitsInfo      = nullptr;
+
+private:
+    BitsState*      m_bitsState     = nullptr;
```
m_bitsState  -> m_pBitsState

amdrexu commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,181 @@
+
+// =====================================================================================================================
+// Helper class for handling Registers defined in SQ_IMG_SAMP_WORD
+class SqImgSampRegHelper : public GfxRegHelper
```
SqImgSampRegHelper -> SqImgSampRegHandler

amdrexu commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,181 @@
+    SqImgSampRegHelper(Builder* pBuilder, Value* pRegister, GfxIpVersion* pGfxIpVersion);
+
+    // Get the current llvm value for the hardware register
+    Value* GetReg(SqSampRegs id);
```
id -> regId

amdrexu commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,181 @@
+    Value* GetReg(SqSampRegs id);
+
+    // Set the current llvm value for the hardware register
+    void SetReg(SqSampRegs id, Value* pParam);
```
id -> regId, pParam -> pRegValue

amdrexu commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,181 @@
+
+// =====================================================================================================================
+// Helper class for handling Registers defined in SQ_IMG_RSRC_WORD
+class SqImgRsrcRegHelper : public GfxRegHelper
```
SqImgRsrcRegHelper -> SqImgRsrcRegHandler

amdrexu commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,181 @@
+    SqImgRsrcRegHelper(Builder* pBuilder, Value* pRegister, GfxIpVersion* pGfxIpVersion);
+
+    // Get the current llvm value for the hardware register
+    Value* GetReg(SqRsrcRegs id);
```
id -> regId

amdrexu commented on lgc/util/llpcGfxRegHelper.h:
```diff
@@ -0,0 +1,181 @@
+    Value* GetReg(SqRsrcRegs id);
+
+    // Set the current llvm value for the hardware register
+    void SetReg(SqRsrcRegs id, Value* pParam);
```
id -> regId, pParam -> pRegValue

amdrexu commented on lgc/util/llpcGfxRegHandler.cpp:
```diff
@@ -25,131 +25,292 @@
+GfxRegHandler::GfxRegHandler(
+    Builder* pBuilder,  // [in] The builder handle
+    Value*   pRegister) // [in] The registered target vec <n x i32>
+:
```
4 space indentation

amdrexu commented on lgc/util/llpcGfxRegHandler.cpp:
```diff
@@ -25,131 +25,292 @@
+    Builder* pBuilder,  // [in] The builder handle
+    Value*   pRegister) // [in] The registered target vec <n x i32>
+:
+GfxRegHandlerBase(pBuilder, pRegister)
```
4 space indentation

amdrexu commented on lgc/util/llpcGfxRegHandler.cpp:
```diff
@@ -25,131 +25,292 @@
 
-    if (auto pVectorTy = dyn_cast<VectorType>(pNewRegister->getType()))
+// =====================================================================================================================
+// Common function for getting the current llvm value for the hardware register
```
current llvm value -> current value

amdrexu commented on lgc/util/llpcGfxRegHandler.cpp:
```diff
@@ -25,131 +25,292 @@
+    Builder*      pBuilder,      // [in] Bound builder context
+    Value*        pRegister,     // [in] Bound register vec <n x i32>
+    GfxIpVersion* pGfxIpVersion) // [in] Current gfxIp version
+:
```
4 space indentation

amdrexu commented on lgc/util/llpcGfxRegHandler.cpp:
```diff
@@ -25,131 +25,292 @@
+    Value*        pRegister,     // [in] Bound register vec <n x i32>
+    GfxIpVersion* pGfxIpVersion) // [in] Current gfxIp version
+:
+GfxRegHandler(pBuilder, pRegister)
```
4 space indentation

amdrexu commented on lgc/util/llpcGfxRegHandler.cpp:
```diff
@@ -25,131 +25,292 @@
+}
+
+// =====================================================================================================================
+// Get the current llvm value for the hardware register
```
current llvm value -> current value

amdrexu commented on lgc/util/llpcGfxRegHandler.h:
```diff
@@ -25,104 +25,157 @@
-    // that are marked as dirty.
-    Value* m_pRegister;
+// SqImgRsrcRegisters ID
+// Merged registers index regarding to g_sqImgRsrcRegBitsGfx9 and g_sqImgRsrcRegBitsGfx10
```
g_sqImgRsrcRegBitsGfx9 -> SqImgRsrcRegBitsGfx9 
g_sqImgRsrcRegBitsGfx10 -> SqImgRsrcRegBitsGfx10

amdrexu commented on lgc/util/llpcGfxRegHandler.h:
```diff
@@ -25,104 +25,157 @@
-    // Set data to a range of bits in indexed DWORD according to BitsInfo
-    void SetBits(const BitsInfo& bitsInfo, Value* pNewBits);
+// SqImgSampRegisters ID
+// Corresponds to g_sqImgSampRegBitsGfx9
```
g_sqImgSampRegBitsGfx9 -> SqImgSampRegBitsGfx9

amdrexu commented on lgc/util/llpcGfxRegHandler.cpp:
```diff
@@ -25,131 +25,292 @@
+    Builder*      pBuilder,      // [in] Bound builder context
+    Value*        pRegister,     // [in] Bound register vec <n x i32>
+    GfxIpVersion* pGfxIpVersion) // [in] Target GFX IP version
+:
```
4 space indentaion

amdrexu commented on lgc/util/llpcGfxRegHandler.cpp:
```diff
@@ -25,131 +25,292 @@
+    Value*        pRegister,     // [in] Bound register vec <n x i32>
+    GfxIpVersion* pGfxIpVersion) // [in] Target GFX IP version
+:
+GfxRegHandler(pBuilder, pRegister)
```
4 space indentaion

jiaxie-amd commented:
@amdrexu I've split the change into 2 commit for that the previous diff was a mess due the the filename overlap.

@trenouf I've rebased the code to your https://github.com/GPUOpen-Drivers/llpc/pull/541, please reivew and reapprove, thanks.

amdrexu commented:
@trenouf It seems the PR is blocked by your request (you marked "request changes"). If OK, you need to re-approve it to unblock it.

JacobHeAMD commented:
retest this please

JacobHeAMD commented:
retest this please

jiaxie-amd commented:
retest this please

amdvlk-admin commented:
### Test summary for commit c30e92a1d8e194350834563474277ce4d6fa21a6
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev bb7398b1c7509a30ba4f7d947b5ad5549a4d58db</li>
<li>SPVGEN: dev dc2b146de03ef54f188172d8fa2f9be708cbfaca</li>
<li>XGL: dev ffd39443fec0ce821b7702173eba7184855c7b5e</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 84d753b2c2274922020557981a17d1667378f746</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 955/955 (100.0%)</li>
<li>Failed: 0/955 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103137/103137 (100.0%)</li>
<li>Failed: 0/103137 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103193/103193 (100.0%)</li>
<li>Failed: 0/103193 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103193/103193 (100.0%)</li>
<li>Failed: 0/103193 (0.0%)</li>
</ul></details>


</ul></details>
